File: backend/app/core/urls.py
"""Whether a URL from the catalogue is one a browser can be sent to.

Every URL this app holds comes from JustWatch -- an offer's deep link, a
provider's icon, a poster -- and every one of them ends up in an ``href`` or a
``src`` without anybody looking at it. An unofficial community API is not a
reason to expect hostile strings, but it is not a reason to rule them out
either, and the cost of asking is one function.

Worth being exact about what this is and is not for. React 19 already blocks a
``javascript:`` href: its ``sanitizeURL`` rewrites one to a thrown error, and its
regex tolerates control characters interleaved between the letters, so the
obvious dodges are covered. Modern browsers block top-level ``data:``
navigation. So this is not closing a live hole, and pretending otherwise would
be the kind of claim that stops a reader trusting the rest of the comments here.

What it does is stop the property depending on somebody else's internals.
React's own message calls its check "a security precaution" rather than a
guarantee, the API is published to more than this one client, and a value that
reaches the database is a value that also reaches log lines and any future
consumer. Holding the rule here costs a scheme comparison.

Pure: no network, no clock, nothing looked up. It answers about the string.
"""

from urllib.parse import urlparse

#: The only two schemes that name somewhere a browser can be sent. Everything
#: else is either script (``javascript:``), a document inlined into the link
#: itself (``data:``), a local resource (``file:``), or an address for something
#: that is not the web at all (``mailto:``, ``tel:``).
_FOLLOWABLE_SCHEMES = frozenset({"http", "https"})

#: ``https:///path`` and friends: a scheme, then no authority at all. RFC
#: parsing calls the host empty; the browser calls the first path segment the
#: host. See the comment in :func:`is_web_url`.
#:
#: Redundant for this module taken alone -- an empty netloc is refused below
#: anyway, and no test here can tell the difference. It earns its place in the
#: frontend mirror, where ``new URL`` reads ``https:///path`` as host ``path``
#: and would otherwise accept it, and it is written on both sides so the two
#: stay readable as the same rule.
_EMPTY_AUTHORITY = ":///"


def is_web_url(value: str | None) -> bool:
    """Whether ``value`` is an ordinary http(s) address with somewhere to go.

    Deliberately a predicate rather than a sanitiser. A rewritten URL is one
    nobody chose: either it still works, and the rewrite bought nothing, or it
    quietly points somewhere other than intended. The caller drops what this
    refuses, and every place the frontend renders one of these already has a
    state for its absence -- a plain label instead of a link, a placeholder
    instead of a poster.
    """
    if not value:
        return False

    # Checked on the string as it arrived, before parsing, because `urlparse`
    # follows the URL standard in stripping tabs, newlines and leading control
    # characters -- so it reports a perfectly ordinary scheme and host for a
    # value that still contains them. The string is what gets stored and what
    # gets printed into a log line, and neither wants a newline in it.
    if any(
        character.isspace() or ord(character) < 0x20 or ord(character) == 0x7F
        for character in value
    ):
        return False

    # Where the two URL standards stop agreeing. `urlparse` implements RFC 3986;
    # the browser that will follow this implements WHATWG, and on a backslash or
    # an empty authority the two do not merely differ in strictness -- they
    # disagree about which host the URL names. Measured:
    #
    #   https://good.test\@evil.test/   RFC: good.test\@evil.test   WHATWG: good.test
    #   https:/\evil.test/              RFC: (no host)              WHATWG: evil.test
    #   https:///path                   RFC: (no host)              WHATWG: path
    #
    # Whatever is decided here, the browser is what navigates, so neither
    # reading can be treated as the truth. Both shapes are refused instead. It
    # costs nothing real -- a URL that genuinely wants a backslash spells it
    # `%5C`, and an empty authority names nowhere under either standard -- and it
    # keeps this check and the browser talking about the same host, which is
    # what a host allowlist would have to rely on.
    if "\\" in value or _EMPTY_AUTHORITY in value:
        return False

    try:
        parsed = urlparse(value)
    except ValueError:
        # A malformed IPv6 host, most plausibly. Unparseable is not followable.
        return False

    # Both halves, and neither is implied by the other. A missing host does not
    # save you from a bad scheme: `javascript://x.test/%0aalert(1)` has a host,
    # and reads to a JavaScript parser as a comment followed by a newline
    # followed by the payload. A good scheme does not save you from a missing
    # host either: an `href` of "https://" is not a link.
    #
    # `.lower()` is belt and braces. `urlparse` already normalises the scheme,
    # so a mutation that removes this changes nothing and no test can tell --
    # it is here so the rule does not silently depend on that staying true.
    return parsed.scheme.lower() in _FOLLOWABLE_SCHEMES and bool(parsed.netloc)


#: Where every poster and provider icon in this app comes from.
#:
#: Not a guess about JustWatch's infrastructure -- it is a constant inside the
#: client library, which builds both fields by concatenating it onto a path out
#: of the API response. ``tests/test_urls.py`` reads the library's own copy and
#: compares, so if they move it, one test says so with a sentence rather than
#: every image quietly failing to load.
#:
#: ``frontend/next.config.ts`` names this same host to Next's image optimiser,
#: and imports it from ``lib/urls.ts`` so there is one spelling of it per side
#: rather than three in total.
CATALOGUE_IMAGE_HOST = "images.justwatch.com"


def is_catalogue_image_url(value: str | None) -> bool:
    """Whether ``value`` is an image this app is willing to make the browser fetch.

    Stricter than :func:`is_web_url`, and deliberately so: an ``href`` waits for
    a click, but a ``src`` is fetched as the page renders, which hands the host
    the viewer's IP address and user agent for nothing. Given what this app is
    about, the request itself is the disclosure.

    The host is affordable to pin because it is not JustWatch's to vary. The
    client library builds these two fields as ``_IMAGES_URL + path``, so any
    other authority means the path did the moving -- and with no separator
    between the two, a field of ``@evil.test/x.jpg`` yields a URL whose host is
    ``evil.test`` and whose scheme and shape are otherwise perfect.

    Compared against the whole authority rather than the parsed hostname, which
    refuses a port and any userinfo in the same comparison. Nothing legitimate
    here carries either, and an authority that is exactly the host is one this
    check and the browser cannot read differently.
    """
    if not is_web_url(value):
        return False
    # `is_web_url` returning True means this parsed once already, so it cannot
    # raise here. Reparsed rather than plumbed through, because a function that
    # returns a bool and a parse result is two functions.
    parsed = urlparse(value)
    # https only. `is_web_url` allows either scheme, since an http link is a
    # link; the optimiser in next.config.ts allows only https, and every real
    # URL from this host is https, so this matches the stricter of the two.
    if parsed.scheme.lower() != "https":
        return False
    # `.lower()` is load-bearing here, unlike on the scheme: `urlparse`
    # normalises the case of `.hostname` but leaves `.netloc` as it arrived.
    return parsed.netloc.lower() == CATALOGUE_IMAGE_HOST

File: frontend/lib/urls.ts
/**
 * Whether a URL from the API is one this app will put in an `href`.
 *
 * **This mirrors `app/core/urls.py` by hand, the way `types.ts` mirrors the
 * Pydantic schemas. Change one, change the other.**
 *
 * The backend already drops an unusable link twice over -- once as it arrives
 * from JustWatch, once on the way back out -- so in a healthy deployment this
 * never fires. It is here for the two cases that outlive a backend rule: rows
 * cached before the rule existed, and the plain fact that a client rendering
 * somebody else's strings into the DOM should not have to assume the server
 * got it right.
 *
 * Worth being straight about the size of this. React 19 already rewrites a
 * `javascript:` href into a thrown error, and its regex tolerates control
 * characters interleaved between the letters, so the obvious dodges are
 * covered; browsers block top-level `data:` navigation on their own. Nothing
 * here is closing a live hole. What it does is stop the property depending on
 * a framework internal that React's own message calls "a security precaution"
 * rather than a guarantee.
 */

// The only two schemes that name somewhere a browser can be sent. `URL`
// lowercases the protocol, so a mixed-case attempt lands here as itself.
const FOLLOWABLE = new Set(["http:", "https:"]);

// Whitespace and the C0/DEL controls, by code point rather than by a character
// class holding the characters themselves -- a regex literal with a raw NUL in
// it is a thing editors and diffs mangle silently.
//
// Checked on the string as it arrived, because `URL` follows the standard in
// stripping tabs and newlines: it reports an ordinary protocol and host for a
// value that still contains them.
function hasControlOrSpace(value: string): boolean {
  for (const character of value) {
    const code = character.codePointAt(0) ?? 0;
    if (code <= 0x20 || code === 0x7f || /\s/.test(character)) return true;
  }
  return false;
}

/**
 * The URL if it can be followed, otherwise null.
 *
 * Returns the original string rather than a normalized one. A rewritten URL is
 * one nobody chose: either it still works and the rewrite bought nothing, or it
 * quietly points somewhere else. Callers already have a state for null -- a
 * plain label instead of a link.
 */
export function webUrl(value: string | null | undefined): string | null {
  if (!value || hasControlOrSpace(value)) return null;

  // Where the two URL standards stop agreeing. The backend's `urlparse`
  // implements RFC 3986 and `new URL` here implements WHATWG, and on a
  // backslash or an empty authority they disagree about which host the URL
  // names -- `https://good.test\@evil.test/` is host `good.test\@evil.test` to
  // one and `good.test` to the other. Both shapes are refused on both sides so
  // that the mirror stays a mirror. A URL that genuinely wants a backslash
  // spells it `%5C`.
  if (value.includes("\\") || value.includes(":///")) return null;

  let parsed: URL;
  try {
    parsed = new URL(value);
  } catch {
    // Relative, protocol-relative, or not a URL at all. `new URL` needs a base
    // for those, and deliberately is not given one: a link that inherits this
    // app's origin is not the third-party destination an offer is supposed to
    // be, and quietly resolving it against the current page would hide that.
    return null;
  }

  // The host as well as the scheme. "https://" parses and names nowhere.
  return FOLLOWABLE.has(parsed.protocol) && parsed.host ? value : null;
}

/**
 * The one host every poster and provider icon comes from.
 *
 * Exported because `next.config.ts` needs the same string for the image
 * optimiser's `remotePatterns`, and a hostname spelled twice on one side is a
 * hostname that will be changed once. `app/core/urls.py` holds the third copy,
 * unavoidably, and its comment says so.
 */
export const CATALOGUE_IMAGE_HOST = "images.justwatch.com";

/**
 * An image URL if it is one this app will let the browser fetch, otherwise null.
 *
 * Stricter than `webUrl`, because an `href` and a `src` are not the same risk.
 * A link waits for a click; an image is requested as the page draws, so a
 * hostile host is handed the viewer's IP address and user agent with no action
 * on their part at all. Given that this app's whole subject is what somebody
 * watches, the request is itself the disclosure.
 *
 * Pinning the host is affordable because it is not really JustWatch's to vary:
 * the client library builds both fields by concatenating its own constant onto
 * a path from the response, so any other authority means the path moved it --
 * and with nothing between the two, a path of `@evil.test/x.jpg` does exactly
 * that while leaving a URL that otherwise looks perfectly ordinary.
 *
 * The two posters go through `next/image`, whose optimiser already refuses an
 * off-list host with a 400 before making any request, so for those this is the
 * second of two checks. The provider icon is a deliberate plain `<img>` -- a
 * fixed 32px logo has nothing for the optimiser to save -- and for that one
 * this is the only check there is.
 */
export function imageUrl(value: string | null | undefined): string | null {
  if (!value || !webUrl(value)) return null;

  // Reparsed rather than threaded out of `webUrl`, which returns the string it
  // was given: a function that answers one question is worth more than one that
  // hands back its working.
  const parsed = new URL(value);

  // `host` is the host the browser will actually connect to: userinfo resolved
  // away, a default port dropped, an IDNA hostname already folded. That makes
  // this the check that matches what happens, which is why it is the one used
  // here rather than a comparison against the raw authority.
  //
  // It also makes this side very slightly more permissive than the backend, and
  // running both over the same table -- 62 cases -- found exactly where:
  //
  //     https://images.justwatch.com:443/x.jpg        backend no, here yes
  //     https://user:pw@images.justwatch.com/x.jpg    backend no, here yes
  //
  // Both of those genuinely do fetch from the allowlisted host, so neither is a
  // hole; `urlparse` compares the authority as a string and so keeps the port
  // and the userinfo, which is stricter than it needs to be rather than wrong.
  // The divergence only ever runs this way, and the backend refuses both on the
  // way in and on the way out, so nothing shaped like this reaches a browser
  // through this app's API at all.
  return parsed.protocol === "https:" && parsed.host === CATALOGUE_IMAGE_HOST
    ? value
    : null;
}
